contracts/hooks/EIP4337Hook.sol

```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.18;
+
+import {IEIP4337Hook, IAccount} from './interfaces/IEIP4337Hook.sol';
+import {IERC1271Wallet} from '../interfaces/IERC1271Wallet.sol';
+import {IModuleCalls} from '../modules/commons/interfaces/IModuleCalls.sol';
+import {ModuleNonce} from '../modules/commons/ModuleNonce.sol';
+import {LibOptim} from '../utils/LibOptim.sol';
+
+contract EIP4337Hook is IEIP4337Hook, ModuleNonce {
+  address public immutable entrypoint;
+  uint256 private constant SIG_VALIDATION_FAILED = 1;
+  bytes4 private constant ERC1271_SELECTOR = 0x1626ba7e;
+
+  /**
+   * Create the EIP-4337 hook for the given entrypoint.
+   */
+  constructor(address _entrypoint) {
+    entrypoint = _entrypoint;
+  }
+
+  modifier onlyEntrypoint() {
+    if (msg.sender != entrypoint) {
+      revert InvalidCaller();
+    }
+    _;
+  }
+
+  /**
+   * Allow the EIP-4337 entrypoint to execute a transaction on the wallet.
+   * @dev This function does not validate as the entrypoint is trusted to have called validateUserOp.
+   * @notice This functions is only callable by the Entrypoint.
+   */
+  function eip4337SelfExecute(IModuleCalls.Transaction[] calldata txs) external payable onlyEntrypoint {
+    // Self execute
+    (bool success, ) = payable(address(this)).call{value: msg.value}(
+      abi.encodeWithSelector(IModuleCalls.selfExecute.selector, txs)
+    );
+    if (!success) {
+      // Bubble up revert reason
+      bytes memory reason = LibOptim.returnData();
+      assembly {
+        revert(add(reason, 0x20), mload(reason))
+      }
+    }
+  }
+
+  /**
+   * Validate and pay for user op.
+   * @dev This must be called by the entrypoint.
+   * @dev GAS opcode is banned during this call, thus max uint256 is used.
+   */
+  function validateUserOp(
+    IAccount.UserOperation calldata userOp,
+    bytes32 userOpHash,
+    uint256 missingAccountFunds
+  ) external onlyEntrypoint returns (uint256 validationData) {
+    // Check nonce.
+    // Note Sequence space encoding is diff to EIP-4337 encoding.
+    _validateNonce(userOp.nonce);
```

Comment:
I understand it will always default to the space 0, correct?

Reply:
Not always but most likely. Here's my research. 

4337 uses a 192-bit “key” and 64-bit “sequence”. Sequence has 160-bit "space" and 96-bit "nonce". 
The space/key is decided by the client. 

If the key is 0 then the space is 0 and things work so long as the Sequence wallet hasn't incremented the nonce yet (unlikely). 
If the key is non zero then the Sequence encoding will shift the key bits into the nonce causing a super high value. 
This breaks things as the nonce is not sequential. 

```js
const decodeNonce = (rawNonce, nonceBits) => ({
    _space: rawNonce >> BigInt(nonceBits),
    _nonce: rawNonce & ((1n << BigInt(nonceBits)) - 1n)
});

rawNonce = BigInt("0x10000000000000002"); // key 1, sequence 2
decodeNonce(rawNonce, 64);
// Object { _space: 1n, _nonce: 2n }
decodeNonce(rawNonce, 96);
// Object { _space: 0n, _nonce: 18446744073709551618n }
```

I think we should use a separate nonce storage for 4337 calls so that parity with the entrypoint is maintained.

Reply:
Or we remove nonce validation from the hook altogether... Not sure why I didn't see that before.

Reply:
Or we ignore it and leave nonce validation to the entrypoint.

Reply:
After discussion, we're opting to ignore it